File: openvox/modules/roles/files/services/healthreport/app/severity.py
"""Deterministic severity classification.

Every observation gets its severity from rules.yml, before the LLM runs. The
model is handed already-classified data and is not permitted to change it -
see llm.py, which strips severity-like keys from the response.
"""

import datetime
from typing import Dict, List, Optional

import yaml

from .model import Observation

FALSE_STRINGS = {"false", "0", "no", "none", "null", ""}


def _age_days(first_seen: Optional[str], now: Optional[datetime.datetime]) -> Optional[float]:
    if not first_seen or now is None:
        return None
    try:
        seen_at = datetime.datetime.fromisoformat(first_seen)
    except ValueError:
        return None
    if seen_at.tzinfo is None and now.tzinfo is not None:
        seen_at = seen_at.replace(tzinfo=now.tzinfo)
    elif seen_at.tzinfo is not None and now.tzinfo is None:
        now = now.replace(tzinfo=seen_at.tzinfo)
    return (now - seen_at).total_seconds() / 86400.0


def load_rules(path: str) -> Dict:
    with open(path, "r") as fh:
        return yaml.safe_load(fh) or {}


def _numeric(value):
    if isinstance(value, bool):
        return None
    try:
        return float(value)
    except (TypeError, ValueError):
        return None


def _truthy(value) -> bool:
    if isinstance(value, bool):
        return value
    if value is None:
        return False
    if isinstance(value, (int, float)):
        return value != 0
    return str(value).strip().lower() not in FALSE_STRINGS


def classify(observation: Observation, rule: Dict, is_new: bool = False,
             now: Optional[datetime.datetime] = None) -> str:
    """Return the severity for one observation. Unknown kinds stay info."""
    if not rule:
        return "info"

    subjects = rule.get("subjects")
    if subjects and observation.subject not in subjects:
        return "info"

    rule_type = rule.get("type")

    if rule_type == "threshold":
        value = _numeric(observation.value)
        if value is None:
            return "info"
        direction = rule.get("direction", "above")
        min_age_days = rule.get("min_age_days")
        for level in ("crit", "warn"):
            limit = rule.get(level)
            if limit is None:
                continue
            limit = float(limit)
            breached = (direction == "above" and value > limit) or \
                       (direction == "below" and value < limit)
            if not breached:
                continue
            if min_age_days is not None:
                # A day-one advisory list is normal; only escalate once it
                # has genuinely sat unresolved (e.g. past the weekly
                # auto-update workflow's own cadence). A brand new
                # observation with no recorded first_seen yet reads as
                # age 0, so it stays info on its first sighting too.
                age = _age_days(observation.first_seen, now)
                if age is None or age < float(min_age_days):
                    return "info"
            observation.threshold = limit
            return level
        return "info"

    if rule_type == "threshold_or_spike":
        # Two independent ways to escalate, whichever fires first.
        #
        # The absolute threshold is deliberately NOT raised to compensate for
        # the spike rule: this widens coverage rather than trading one signal
        # for another. A container that is loud every day keeps reporting, and
        # one that quietly triples while staying under the line now reports too.
        value = _numeric(observation.value)
        if value is None:
            return "info"

        direction = rule.get("direction", "above")
        for level in ("crit", "warn"):
            limit = rule.get(level)
            if limit is None:
                continue
            limit = float(limit)
            if (direction == "above" and value > limit) or \
               (direction == "below" and value < limit):
                observation.threshold = limit
                return level

        previous = _numeric(observation.previous_value)
        factor = float(rule.get("spike_factor", 3.0))
        floor = float(rule.get("spike_floor", 0))
        # A first sighting has no baseline and must not read as a spike, and a
        # previous value of zero would make any growth infinite.
        if previous and previous > 0 and value >= floor and value >= previous * factor:
            # Say why this escalated. Without it the reader sees a number well
            # under the stated threshold and no reason for the warning.
            # (classify already mutates `threshold`, so annotating here is
            # consistent with how the absolute path reports itself.)
            observation.message = "%s — up %.1fx from %s" % (
                observation.message, value / previous, f"{previous:,.0f}",
            )
            return rule.get("spike_level", "warn")

        return "info"

    if rule_type == "equals":
        value = str(observation.value)
        if value in [str(v) for v in rule.get("crit_values", [])]:
            return "crit"
        if value in [str(v) for v in rule.get("warn_values", [])]:
            return "warn"
        return "info"

    if rule_type == "not_equals":
        ok = [str(v) for v in rule.get("ok_values", [])]
        if str(observation.value) not in ok:
            return rule.get("level", "warn")
        return "info"

    if rule_type == "truthy":
        return rule.get("level", "warn") if _truthy(observation.value) else "info"

    if rule_type == "falsy":
        return rule.get("level", "warn") if not _truthy(observation.value) else "info"

    if rule_type == "present":
        return rule.get("level", "warn")

    if rule_type == "new_only":
        # Only interesting the first time it is seen. Steady-state noise stays
        # info so it never reaches the report body.
        return rule.get("level", "warn") if is_new else "info"

    return "info"


def apply(observations: List[Observation], rules: Dict, new_ids=frozenset(),
          now: Optional[datetime.datetime] = None) -> List[Observation]:
    table = (rules or {}).get("rules", {})
    for observation in observations:
        observation.severity = classify(
            observation,
            table.get(observation.kind, {}),
            is_new=observation.id in new_ids,
            now=now,
        )
    return observations

File: tools/internal/healthreport/severity.go
package healthreport

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"gopkg.in/yaml.v3"
)

var falseStrings = map[string]bool{"false": true, "0": true, "no": true, "none": true, "null": true, "": true}

// Rule is one entry from rules.yml. Severity is decided here, deterministically,
// before the LLM is ever called - see llm.go.
type Rule struct {
	Type        string   `yaml:"type"`
	Direction   string   `yaml:"direction"`
	Warn        *float64 `yaml:"warn"`
	Crit        *float64 `yaml:"crit"`
	MinAgeDays  *float64 `yaml:"min_age_days"`
	SpikeFactor *float64 `yaml:"spike_factor"`
	SpikeFloor  *float64 `yaml:"spike_floor"`
	SpikeLevel  string   `yaml:"spike_level"`
	CritValues  []any    `yaml:"crit_values"`
	WarnValues  []any    `yaml:"warn_values"`
	OkValues    []any    `yaml:"ok_values"`
	Level       string   `yaml:"level"`
	Subjects    []string `yaml:"subjects"`
}

type RulesFile struct {
	Rules map[string]Rule `yaml:"rules"`
}

func LoadRules(path string) (RulesFile, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return RulesFile{}, err
	}
	var rf RulesFile
	if err := yaml.Unmarshal(raw, &rf); err != nil {
		return RulesFile{}, err
	}
	return rf, nil
}

func numeric(value any) (float64, bool) {
	switch v := value.(type) {
	case bool:
		return 0, false
	case float64:
		return v, true
	case float32:
		return float64(v), true
	case int:
		return float64(v), true
	case int64:
		return float64(v), true
	case string:
		f, err := strconv.ParseFloat(v, 64)
		if err != nil {
			return 0, false
		}
		return f, true
	default:
		return 0, false
	}
}

func truthy(value any) bool {
	switch v := value.(type) {
	case bool:
		return v
	case nil:
		return false
	case float64:
		return v != 0
	case float32:
		return v != 0
	case int:
		return v != 0
	case int64:
		return v != 0
	default:
		return !falseStrings[strings.ToLower(strings.TrimSpace(fmt.Sprint(v)))]
	}
}

func ageDays(firstSeen any, now *time.Time) (float64, bool) {
	s, ok := firstSeen.(string)
	if !ok || s == "" || now == nil {
		return 0, false
	}
	seenAt, err := parseISOTime(s)
	if err != nil {
		return 0, false
	}
	return now.Sub(seenAt).Hours() / 24.0, true
}

func parseISOTime(s string) (time.Time, error) {
	for _, layout := range []string{time.RFC3339Nano, time.RFC3339, "2006-01-02T15:04:05.999999"} {
		if t, err := time.Parse(layout, s); err == nil {
			return t, nil
		}
	}
	return time.Time{}, fmt.Errorf("unparseable time: %s", s)
}

func valuesContain(values []any, want string) bool {
	for _, v := range values {
		if fmt.Sprint(v) == want {
			return true
		}
	}
	return false
}

// Classify returns the severity for one observation. Unknown kinds stay info.
func Classify(obs *Observation, rule Rule, isNew bool, now *time.Time) string {
	if rule.Type == "" {
		return "info"
	}

	if len(rule.Subjects) > 0 {
		found := false
		for _, s := range rule.Subjects {
			if s == obs.Subject {
				found = true
				break
			}
		}
		if !found {
			return "info"
		}
	}

	switch rule.Type {
	case "threshold":
		value, ok := numeric(obs.Value)
		if !ok {
			return "info"
		}
		direction := rule.Direction
		if direction == "" {
			direction = "above"
		}
		for _, level := range []string{"crit", "warn"} {
			limit := rule.Crit
			if level == "warn" {
				limit = rule.Warn
			}
			if limit == nil {
				continue
			}
			breached := (direction == "above" && value > *limit) || (direction == "below" && value < *limit)
			if !breached {
				continue
			}
			if rule.MinAgeDays != nil {
				// A day-one advisory list is normal; only escalate once it
				// has genuinely sat unresolved. A brand new observation
				// with no recorded first_seen yet reads as age 0, so it
				// stays info on its first sighting too.
				age, ok := ageDays(obs.FirstSeen, now)
				if !ok || age < *rule.MinAgeDays {
					return "info"
				}
			}
			obs.Threshold = *limit
			return level
		}
		return "info"

	case "threshold_or_spike":
		// Two independent ways to escalate, whichever fires first. The
		// absolute threshold is deliberately NOT raised to compensate for
		// the spike rule: this widens coverage rather than trading one
		// signal for another.
		value, ok := numeric(obs.Value)
		if !ok {
			return "info"
		}
		direction := rule.Direction
		if direction == "" {
			direction = "above"
		}
		for _, level := range []string{"crit", "warn"} {
			limit := rule.Crit
			if level == "warn" {
				limit = rule.Warn
			}
			if limit == nil {
				continue
			}
			if (direction == "above" && value > *limit) || (direction == "below" && value < *limit) {
				obs.Threshold = *limit
				return level
			}
		}

		previous, hasPrevious := numeric(obs.PreviousValue)
		factor := 3.0
		if rule.SpikeFactor != nil {
			factor = *rule.SpikeFactor
		}
		floor := 0.0
		if rule.SpikeFloor != nil {
			floor = *rule.SpikeFloor
		}
		// A first sighting has no baseline and must not read as a spike,
		// and a previous value of zero would make any growth infinite.
		if hasPrevious && previous > 0 && value >= floor && value >= previous*factor {
			obs.Message = fmt.Sprintf("%s — up %.1fx from %s", obs.Message, value/previous, formatThousands(previous))
			level := rule.SpikeLevel
			if level == "" {
				level = "warn"
			}
			return level
		}
		return "info"

	case "equals":
		value := fmt.Sprint(obs.Value)
		if valuesContain(rule.CritValues, value) {
			return "crit"
		}
		if valuesContain(rule.WarnValues, value) {
			return "warn"
		}
		return "info"

	case "not_equals":
		value := fmt.Sprint(obs.Value)
		if !valuesContain(rule.OkValues, value) {
			level := rule.Level
			if level == "" {
				level = "warn"
			}
			return level
		}
		return "info"

	case "truthy":
		level := rule.Level
		if level == "" {
			level = "warn"
		}
		if truthy(obs.Value) {
			return level
		}
		return "info"

	case "falsy":
		level := rule.Level
		if level == "" {
			level = "warn"
		}
		if !truthy(obs.Value) {
			return level
		}
		return "info"

	case "present":
		level := rule.Level
		if level == "" {
			level = "warn"
		}
		return level

	case "new_only":
		// Only interesting the first time it is seen. Steady-state noise
		// stays info so it never reaches the report body.
		if isNew {
			level := rule.Level
			if level == "" {
				level = "warn"
			}
			return level
		}
		return "info"
	}

	return "info"
}

func formatThousands(v float64) string {
	// Matches Python's f"{v:,.0f}".
	s := strconv.FormatFloat(v, 'f', 0, 64)
	neg := strings.HasPrefix(s, "-")
	if neg {
		s = s[1:]
	}
	var out []byte
	for i, c := range []byte(s) {
		if i > 0 && (len(s)-i)%3 == 0 {
			out = append(out, ',')
		}
		out = append(out, c)
	}
	if neg {
		return "-" + string(out)
	}
	return string(out)
}

// Apply classifies every observation in place.
func Apply(observations []*Observation, rules RulesFile, newIDs map[string]bool, now *time.Time) {
	for _, obs := range observations {
		rule := rules.Rules[obs.Kind]
		obs.Severity = Classify(obs, rule, newIDs[obs.ID], now)
	}
}
